lib/helpers.py

```diff
@@ -0,0 +1,389 @@
+        rev_per_conversion_test = test_revenue / test_conversions
+    if control_conversions > 0:
+        rev_per_conversion_control = control_revenue / control_conversions
+
+    test_cvr = test_conversions / test_group_size
+    control_cvr = control_conversions / control_group_size
+
+    uplift = 0
+    if control_cvr > 0:
+        uplift = test_cvr / control_cvr - 1
+
+    # calculate statistical significance
+    control_successes, test_successes = control_conversions, test_conversions
+    if use_converters_for_significance or max(test_cvr, control_cvr) > 1.0:
+        control_successes, test_successes = control_converters, test_converters
+    chi_df = pd.DataFrame({
+        "conversions": [control_successes, test_successes],
+        "total": [control_group_size, test_group_size]
+    }, index=['control', 'test'])
+    # CHI square calculation will fail with insufficient data
+    # Fallback to no significance
+    try:
+        chi, p, *_ = scipy.stats.chi2_contingency(
+            pd.concat([chi_df.total - chi_df.conversions, chi_df.conversions], axis=1), correction=False)
+    except:
+        chi, p = 0, 1.0
+        
+    # bonferroni correction with equal weights - if we have multiple hypothesis:
+    # https://en.wikipedia.org/wiki/Bonferroni_correction
+    significant = p < 0.05/m_hypothesis
+
+    dataframe_dict = {
+        "ad spend": ad_spend,
+        "total revenue": test_revenue + control_revenue,
+        "test group size": test_group_size,
+        "test conversions": test_conversions,
+        "test converters": test_converters,
+        "test revenue": test_revenue,
+        "control group size": control_group_size,
+        "control conversions": control_conversions,
+        "control_converters": control_converters,
+        "control revenue": control_revenue,
+        "ratio test/control": ratio,
+        "control conversions (scaled)": scaled_control_conversions,
+        "control revenue (scaled)": scaled_control_revenue_micros / 10 ** 6,
+        "incremental conversions": incremental_conversions,
+        "incremental converters": incremental_converters,
+        "incremental revenue": incremental_revenue,
+        "rev/conversions test": rev_per_conversion_test,
+        "rev/conversions control": rev_per_conversion_control,
+        "test CVR": test_cvr,
+        "control CVR": control_cvr,
+        "CVR Uplift": uplift,
+        "iROAS": iroas,
+        "cost per incr. converter": cost_per_incremental_converter,
+        "iCPA": icpa,
+        "chi^2": chi,
+        "p-value": p,
+        "significant": significant
+    }
+
+    # show results as a dataframe
+    return pd.DataFrame(
+        dataframe_dict,
+        index=[index_name],
+    ).transpose()
+
+"""### Calculate and display uplift report for the data set as a whole
+This takes the whole data set and calculates uplift KPIs.
+"""
+
+def uplift_report(bids_df, attributions_df, groups, per_campaign_results, use_converters_for_significance):
```

Comment:
(low) Both `uplift` and `uplift_report` just ask to be converted into one class.

Reply:
...and a bunch of other things :)